components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/manager/FxaAccountManager.kt

```diff
@@ -924,53 +776,71 @@ open class FxaAccountManager(
             )
         }
     }
 
-    private suspend fun doAuthenticate(): Event? {
-        val authFlowUrl = account.beginOAuthFlowAsync(scopes).await()
-        if (authFlowUrl == null) {
-            oauthObservers.notifyObservers { onError() }
-            return Event.FailedToAuthenticate
-        }
-        latestAuthState = authFlowUrl.state
-        oauthObservers.notifyObservers { onBeginOAuthFlow(authFlowUrl) }
-        return null
-    }
-
-    private suspend fun postAuthenticated(authType: AuthType, declinedEngines: Set<SyncEngine>? = null) {
+    private fun persistDeclinedEngines(declinedEngines: Set<SyncEngine>) {
         // Sync may not be configured at all (e.g. won't run), but if we received a
         // list of declined engines, that indicates user intent, so we preserve it
         // within SyncEnginesStorage regardless. If sync becomes enabled later on,
         // we will be respecting user choice around what to sync.
-        declinedEngines?.let {
-            val enginesStorage = SyncEnginesStorage(context)
-            declinedEngines.forEach { declinedEngine ->
-                enginesStorage.setStatus(declinedEngine, false)
-            }
+        val enginesStorage = SyncEnginesStorage(context)
+        declinedEngines.forEach { declinedEngine ->
+            enginesStorage.setStatus(declinedEngine, false)
+        }
 
-            // Enable all engines that were not explicitly disabled. Only do this in
-            // the presence of a "declinedEngines" list, since that indicates user
-            // intent. Absence of that list means that user was not presented with a
-            // choice during authentication, and so we can't assume an enabled state
-            // for any of the engines.
-            syncConfig?.supportedEngines?.forEach { supportedEngine ->
-                if (!declinedEngines.contains(supportedEngine)) {
-                    enginesStorage.setStatus(supportedEngine, true)
-                }
+        // Enable all engines that were not explicitly disabled. Only do this in
+        // the presence of a "declinedEngines" list, since that indicates user
+        // intent. Absence of that list means that user was not presented with a
+        // choice during authentication, and so we can't assume an enabled state
+        // for any of the engines.
+        syncConfig?.supportedEngines?.forEach { supportedEngine ->
+            if (!declinedEngines.contains(supportedEngine)) {
+                enginesStorage.setStatus(supportedEngine, true)
             }
         }
+    }
 
-        // Before any sync workers have a chance to access it, make sure our SyncAuthInfo cache is hot.
-        maybeUpdateSyncAuthInfoCache()
+    sealed class RetryResult<out T> {
+        data class Success<out T>(val value: T) : RetryResult<T>()
+        object Failure : RetryResult<Nothing>()
+    }
+
+    private suspend fun <T> withRetries(retryCount: Int, block: suspend () -> T): RetryResult<T> {
```

Comment:
Nice! Let's add a unit test for this one though. It could potentially also live outside this module, if we want it later.